### ISSUE TITLE
"Transfrom Object..." order of operations is wrong

### ISSUE THREAD

**Issue by peteihis** [ArtOfIllusion/ArtOfIllusion#302]
Hi.

I was actually looking at a different issue, when I realized this. Example:
- Make mesh objects:
- Transform Objects ... -> Move X = 2.0, Rotate Z = 90, Rotate around Center of Selection -> OK

It rotates (and scales) as expected but moves to y-direction, not x as instructed. What seem to happen is that it first performs the move and the rotates around the _old_ center of selection.

There are uses for moving into rotated directions but then the parameters on the UI should clearly state that. I rarely use the different operations of transform simultaneously, so this has not come up.

My plan:
- Add scaling and rotaion opition around **Center of Scene**. _(This was what I was looking at -- Helps to rescale imported scenes to "AoI-comfortable" dimensions. Myself, I used a script to do this to imported parts of a scene, when I needed it more)_
- Add a checkbox for "**Rotate the direction of Move**".

**Comment by peteihis** [comment]
And btw, there is the class "TransformPointsDialog" that is not used in AoI Core code. Shouldn't it be marked deprecated?

**Comment by Sailsman63** [comment]
Experimenting with a single object, I see that:

- Rotation is around Object/Selection Z axis
- Movement is in the direction of "Whatever part of the object used to face Selection X"

This is old code - I'll need to do some looking about to understand it.

It may actually be safe to simply remove the `TransformPointsDialog`, as it doesn't actually *do* anything.

**Comment by makiam** [comment]
Experimenting with a single object, I see that:
Entering rotation 45 degrees in Transform Dialog gives -45 degrees in Object Properties panel

**Comment by peteihis** [comment]
> Entering rotation 45 degrees in Transform Dialog gives -45 degrees in Object Properties panel

That is correct. The transformation tool uses the normal Cartesian math for rotations whereas `CoordinateSystem` turns rotations to the opposite direction. The conversion happens inside 'CoordinateSystem' by axis directions.

Personally I can't begin to imagine why that ***** has been coded into CoordinateSystem in the first place. It is just plain stupid! It messes up rotations, that you produce by scripting (depending on if you used vectors and matrices or angles) and when programming view moves I had to pay extra attention to transfering angles between common math and CoordinateSystem.

I actually checked and tested once, what it would take to flip the CoordinateSystem angles to match the rest of the world and it turned out to be more simple that you might think. There is a limited number of places where angles appear and things like animation tracks could be converted by a pretty simple fixed logic. Flipping an angle has some other effects that are every time the same. Then there are things like the Distribute script that use angles. Distribute would start to random-rotate objects in the opposite way it does now, but of course anything that was distributed and rotated before would not move.

**Comment by Sailsman63** [comment]
>Entering rotation 45 degrees in Transform Dialog gives -45 degrees in Object Properties panel

This will vary depending on the object layout. The transform tool is working in object space, the ObjectProperties panel is in scene space. I'm not sure that this is actually *wrong*, though it is becoming clear that the interface doesn't make that obvious to the user.

**Comment by peteihis** [comment]
> The transform tool is working in object space

No.

When you are moving or rotating objects on the scene, it works in the scene space, using Cartesian coordinates. Only scaling happens in object space. CoordinateSystem angles are still to the oppoosite direction of Cartesian. That is in the CoordinateSystem code and it is deliberate, I just can't se why.

**Comment by peteihis** [comment]
Yes. individual functions of Transform Object work correctly. That was never the problem.

The original issue is, about performing moves and rotation is one step. The expectation would be that the order of operations would be as they appear on the dialog: First move _along the currently active_ coordinate system and then rotate. What now happens is that the Transform Object rotates, not only the objects but also _the coordinate directions of the movement_ and moves to the rotated directions. That might be useful if you know and expect it but that would not be what I'd expect seeing the UI.

> the "top" cube also has 180 degree rotations around Y and Z. That is weird.

That is weird. That is CoordinateSystem at work. I'm not sure where that comes form but somehow CS tends to favor the ±90/180° numbers when it converts swapped axis directions to angles. I have no idea why that happens. (Numerical rounding error, some <> issue....) This was also the case with bones. I seem to remember that some optimizations was added there, because close-to-180 angles limit movements.

But quirks of CoordinateSystem were not in the scope of this issue, the Transfrom too is. I already started to design a modification to the TO-tool, the changes are easy after all... I'm not sure if my plan on the forts message was undestandable but I'd:
 - Make the order of operations user selectable: Either move first rotate then (as you'd expect) or rotate the group first and move to the rotated directions (as it does now).
 - Add one scaling center option, "Scene origin" _(the most obvious option to me)_. This is why I started to look at it in the first place. When I just want to scale the entire scene retaining the relative positions to the coordinates, it is not possible with the current choices. Say, if have imported something in mm-scale and need it in meters (happened to me) and then scale by 0.001, scaling happens about the center of the selection, which is easily several meters off the center.
 
A real life case from a few years back (roughly): I had a group of objects that was 48 meters long. The center of the group was at x = 18 meters. The other software exported it in millimeters, resulting a set 48 000 units in length with it's center at x = 18 000. I needed to scale that down do 48 units in size but after the scaling the center was still at 18 000 units. Of course all the export/import could not be done in one step, so the next parts of the scene would be some 5 meters (5000 units) in size and located at 22 meters. Scale that --> I have two groups of dots and flakes on the scene; one group at x = 18000 and the other at x = 22000. Hmm....

**Comment by peteihis** [comment]
And to visualize my point (the finding):

![Transfrom tool](https://github.com/ArtOfIllusion/ArtOfIllusion/assets/22843957/dbc28bcd-e3e3-470a-b72a-e16d499f0c5e)

**Comment by Sailsman63** [comment]
> to visualize my point 

Okay, *that* should be easy enough to patch. PR welcome. I'd stick with "Move->Rotate->Scale" sequence, as it would be the simplest. Adding more configuration isn't going to make this easier to use - remember, *simple* UI! If someone needs to do things in other orders, they can do two operations.

> Add one scaling center option, "Scene origin"

Maybe both scale & rotate at the same time? Third radio button, disables the "Move" inputs?


I did figure out the swapped 180 degree rotations, I think: Gimbal Lock, which is inevitable when storing/calculating rotations as Euler Angles. This is also what makes bones weird at times. There are mathematical alternatives, such as Quaternions and geometric Rotors, but the UI for these can get tricky.

**Comment by peteihis** [comment]
> Maybe both scale & rotate at the same time?

Yes, the selected origin is, of course, THE oring of the entire set of operations.

> Third radio button, disables the "Move" inputs?

??? Is you don't want to move, you leave the move values 0.0. -- Anyway as long as move happens in the unrotated coordinates, the origin selection is irrelevant. The travel is the same regardless of the origin.

**Comment by peteihis** [comment]
> Gimbal Lock, which is inevitable..

Gimbal Lock is inevitable with certain combinations of angle but it does not explain the tendency to favor combinations of 180°s when 0° s would do the job. I agree that Quaternions would not make sense on the UI.

**Comment by peteihis** [comment]
Just as a side track, I had a look at the coordinate angles thing. Where it goes wrong is in the `findRotationAngles()` of CoordinateSystem, in the first `if`-block, startin on line 227:

~~~~Java
    if (zdir.x == 0.0 && zdir.z == 0.0)
      {
        d = Math.sqrt(updir.x*updir.x+updir.z*updir.z);
        yrot = Math.acos(updir.z/d);
        if (Double.isNaN(yrot))
          yrot = Math.acos(updir.z > 0.0 ? 1.0 : -1.0);
        if (zdir.y > 0.0)
          yrot *= -1.0;
      }
~~~~

This is the Gimbal Lock case when the z-direction of the object aligns with the y-direction of the scene. This works fine when you add and object from Bottom view, but in the Top view case the y-rotation is 180 degrees off and the rest of the method compensates that with the z-rotation. The angles do represent the correct orientation of the object but unnecessarily counter affecting each others.

If I am not mistaken this should get fixed by adding a `Math.PI` somewhere in the last `if`-condition. I'd expect rest of the method to work correctly, but if not the needed changes should be something in the similar level with `yrot`.

**Comment by peteihis** [comment]
> If I am not mistaken

I was (or so it seems). I did not get to testing with the code but there seems to be something more going on...
